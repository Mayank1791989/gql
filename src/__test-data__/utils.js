/* @flow */
import path from 'path';
import { dedent } from 'dentist';
import splitLines from '../shared/splitLines';

export function getSchema() {
  jest.mock('../shared/watch');
  const watch = require('../shared/watch').default;
  const GQLService = require('../').GQLService;

  const gql = new GQLService({ cwd: path.resolve('src/__test-data__/') });
  // $FlowDisableNextLine
  watch.__triggerChange();
  return gql._schemaBuilder.getSchema();
}

export function getDefLocations() {
  return {
    Player: {
      start: { line: 12, column: 1 },
      end: { line: 16, column: 2 },
      path: path.resolve('src/__test-data__/schema/query.gql'),
    },
    Player_id: {
      start: { line: 13, column: 3 },
      end: { line: 13, column: 10 },
      path: path.resolve('src/__test-data__/schema/query.gql'),
    },
    Player_image_arg_size: {
      start: { line: 15, column: 9 },
      end: { line: 15, column: 19 },
      path: path.resolve('src/__test-data__/schema/query.gql'),
    },

    Team: {
      start: { line: 17, column: 1 },
      end: { line: 20, column: 2 },
      path: path.resolve('src/__test-data__/schema/query.gql'),
    },
    Node: {
      start: { line: 8, column: 1 },
      end: { line: 10, column: 2 },
      path: path.resolve('src/__test-data__/schema/query.gql'),
    },
    Edge: {
      start: { line: 27, column: 1 },
      end: { line: 30, column: 2 },
      path: path.resolve('src/__test-data__/schema/query.gql'),
    },

    Mutation: {
      start: { line: 2, column: 1 },
      end: { line: 5, column: 2 },
      path: path.resolve('src/__test-data__/schema/mutation.gql'),
    },
    Mutation_PlayerCreate: {
      start: { line: 4, column: 3 },
      end: { line: 4, column: 63 },
      path: path.resolve('src/__test-data__/schema/mutation.gql'),
    },
    Mutation_PlayerCreateInput: {
      start: { line: 6, column: 1 },
      end: { line: 9, column: 2 },
      path: path.resolve('src/__test-data__/schema/mutation.gql'),
    },
    Mutation_PlayerCreateInput_id: {
      start: { line: 8, column: 3 },
      end: { line: 8, column: 10 },
      path: path.resolve('src/__test-data__/schema/mutation.gql'),
    },

    Query: {
      start: { line: 2, column: 1 },
      end: { line: 6, column: 2 },
      path: path.resolve('src/__test-data__/schema/query.gql'),
    },
    Query_viewer: {
      start: { line: 5, column: 3 },
      end: { line: 5, column: 18 },
      path: path.resolve('src/__test-data__/schema/query.gql'),
    },

    Role: {
      start: { line: 1, column: 1 },
      end: { line: 5, column: 2 },
      path: path.resolve('src/__test-data__/schema/enums.gql'),
    },

    CustomScalar: {
      start: { line: 1, column: 1 },
      end: { line: 1, column: 20 },
      path: path.resolve('src/__test-data__/schema/scalars.gql'),
    },

    customDirective: {
      start: { line: 2, column: 1 },
      end: { line: 7, column: 25 },
      path: path.resolve('src/__test-data__/schema/directives.gql'),
    },
    customDirective_argIf: {
      start: { line: 4, column: 3 },
      end: { line: 4, column: 15 },
      path: path.resolve('src/__test-data__/schema/directives.gql'),
    },
  };
}

export function getRefLocations() {
  return {
    Player: [
      {
        start: { line: 14, column: 11 },
        end: { line: 14, column: 17 },
        path: path.resolve('src/__test-data__/schema/mutation.gql'),
      },
      {
        start: { line: 24, column: 7 },
        end: { line: 24, column: 13 },
        path: path.resolve('src/__test-data__/schema/query.gql'),
      },
      {
        start: { line: 12, column: 6 },
        end: { line: 12, column: 12 },
        path: path.resolve('src/__test-data__/schema/query.gql'),
      },
    ],

    Edge: [
      {
        start: { line: 28, column: 11 },
        end: { line: 28, column: 15 },
        path: path.resolve('src/__test-data__/schema/query.gql'),
      },
    ],

    Role: [
      {
        start: { line: 10, column: 9 },
        end: { line: 10, column: 13 },
        path: path.resolve('src/__test-data__/schema/mutation.gql'),
      },
      {
        start: { line: 34, column: 9 },
        end: { line: 34, column: 13 },
        path: path.resolve('src/__test-data__/schema/query.gql'),
      },
      {
        start: { line: 36, column: 15 },
        end: { line: 36, column: 19 },
        path: path.resolve('src/__test-data__/schema/query.gql'),
      },
      {
        start: { line: 1, column: 6 },
        end: { line: 1, column: 10 },
        path: path.resolve('src/__test-data__/schema/enums.gql'),
      },
    ],

    CustomScalar: [
      {
        start: { line: 35, column: 10 },
        end: { line: 35, column: 22 },
        path: path.resolve('src/__test-data__/schema/query.gql'),
      },
      {
        start: { line: 1, column: 8 },
        end: { line: 1, column: 20 },
        path: path.resolve('src/__test-data__/schema/scalars.gql'),
      },
    ],
  };
}

export function getHints() {
  return {
    ObjectTypes: [
      {
        description: 'Mutation contains all allowed mutations',
        text: 'Mutation',
        type: 'Object',
      },
      {
        description: '',
        text: 'PlayerCreatePayload',
        type: 'Object',
      },
      {
        description: 'Query is the root query object',
        text: 'Query',
        type: 'Object',
      },
      {
        description: '',
        text: 'Player',
        type: 'Object',
      },
      {
        description: '',
        text: 'Team',
        type: 'Object',
      },
      {
        description: '',
        text: 'Viewer',
        type: 'Object',
      },
      {
        description: '',
        text: 'NewPlayer',
        type: 'Object',
      },
    ],

    OutputTypes: [
      {
        description: '',
        text: 'Role',
        type: 'Enum',
      },
      {
        description: 'Mutation contains all allowed mutations',
        text: 'Mutation',
        type: 'Object',
      },
      {
        description: '',
        text: 'PlayerCreatePayload',
        type: 'Object',
      },
      {
        description: 'Query is the root query object',
        text: 'Query',
        type: 'Object',
      },
      {
        description: '',
        text: 'Node',
        type: 'Interface',
      },
      {
        description: '',
        text: 'Player',
        type: 'Object',
      },
      {
        description: '',
        text: 'Team',
        type: 'Object',
      },
      {
        description: '',
        text: 'Viewer',
        type: 'Object',
      },
      {
        description: '',
        text: 'Edge',
        type: 'Interface',
      },
      {
        description: '',
        text: 'NewPlayer',
        type: 'Object',
      },
      {
        description: '',
        text: 'CustomScalar',
        type: 'Scalar',
      },
      {
        description: 'The `String` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.',
        text: 'String',
        type: 'Scalar',
      },
      {
        description: 'The `Int` scalar type represents non-fractional signed whole numeric values. Int can represent values between -(2^31) and 2^31 - 1. ',
        text: 'Int',
        type: 'Scalar',
      },
      {
        description: 'The `Float` scalar type represents signed double-precision fractional values as specified by [IEEE 754](http://en.wikipedia.org/wiki/IEEE_floating_point). ',
        text: 'Float',
        type: 'Scalar',
      },
      {
        description: 'The `Boolean` scalar type represents `true` or `false`.',
        text: 'Boolean',
        type: 'Scalar',
      },
      {
        description: 'The `ID` scalar type represents a unique identifier, often used to refetch an object or as key for a cache. The ID type appears in a JSON response as a String; however, it is not intended to be human-readable. When expected as an input type, any string (such as `"4"`) or integer (such as `4`) input value will be accepted as an ID.',
        text: 'ID',
        type: 'Scalar',
      },
    ],

    CompositeTypes: [
      {
        description: 'Mutation contains all allowed mutations',
        text: 'Mutation',
        type: 'Object',
      },
      {
        description: '',
        text: 'PlayerCreatePayload',
        type: 'Object',
      },
      {
        description: 'Query is the root query object',
        text: 'Query',
        type: 'Object',
      },
      {
        description: '',
        text: 'Node',
        type: 'Interface',
      },
      {
        description: '',
        text: 'Player',
        type: 'Object',
      },
      {
        description: '',
        text: 'Team',
        type: 'Object',
      },
      {
        description: '',
        text: 'Viewer',
        type: 'Object',
      },
      {
        description: '',
        text: 'Edge',
        type: 'Interface',
      },
      {
        description: '',
        text: 'NewPlayer',
        type: 'Object',
      },
    ],

    InputTypes: [
      {
        description: '',
        text: 'Role',
        type: 'Enum',
      },
      {
        description: '',
        text: 'PlayerCreateInput',
        type: 'Input',
      },
      {
        description: '',
        text: 'CustomScalar',
        type: 'Scalar',
      },
      {
        description: 'The `String` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.',
        text: 'String',
        type: 'Scalar',
      },
      {
        description: 'The `Int` scalar type represents non-fractional signed whole numeric values. Int can represent values between -(2^31) and 2^31 - 1. ',
        text: 'Int',
        type: 'Scalar',
      },
      {
        description: 'The `Float` scalar type represents signed double-precision fractional values as specified by [IEEE 754](http://en.wikipedia.org/wiki/IEEE_floating_point). ',
        text: 'Float',
        type: 'Scalar',
      },
      {
        description: 'The `Boolean` scalar type represents `true` or `false`.',
        text: 'Boolean',
        type: 'Scalar',
      },
      {
        description: 'The `ID` scalar type represents a unique identifier, often used to refetch an object or as key for a cache. The ID type appears in a JSON response as a String; however, it is not intended to be human-readable. When expected as an input type, any string (such as `"4"`) or integer (such as `4`) input value will be accepted as an ID.',
        text: 'ID',
        type: 'Scalar',
      },
    ],

    InterfaceTypes: [
      {
        description: '',
        text: 'Node',
        type: 'Interface',
      },
      {
        description: '',
        text: 'Edge',
        type: 'Interface',
      },
    ],

    PlayerTypeFields: [
      {
        description: '',
        text: 'id',
        type: 'ID!',
      },
      {
        description: '',
        text: 'name',
        type: 'String!',
      },
      {
        description: '',
        text: 'image',
        type: 'String!',
      },
    ],

    PlayerTypeImageFieldArgs: [
      {
        description: '',
        text: 'size',
        type: 'Int!',
      },
    ],

    TypesImplementsNode: [
      {
        description: '',
        text: 'Player',
        type: 'Object',
      },
      {
        description: '',
        text: 'Team',
        type: 'Object',
      },
      {
        description: '',
        text: 'Node',
        type: 'Interface',
      },
    ],

    PossibleMutations: [
      {
        description: 'Player Create contains all allowed mutations',
        text: 'PlayerCreate',
        type: 'PlayerCreatePayload',
      },
    ],

    QueryFields: [
      {
        description: '',
        text: 'node',
        type: 'Node',
      },
      {
        description: '',
        text: 'nodes',
        type: '[Node]',
      },
      {
        description: '',
        text: 'viewer',
        type: 'Viewer!',
      },
      {
        description: 'Access the current type schema of this server.',
        text: '__schema',
        type: '__Schema!',
      },
      {
        description: 'Request the type information of a single type.',
        text: '__type',
        type: '__Type',
      },
    ],

    EnumRoleValues: [
      {
        description: '',
        text: 'roleA',
        type: 'Role',
      },
      {
        description: '',
        text: 'roleB',
        type: 'Role',
      },
      {
        description: '',
        text: 'roleC',
        type: 'Role',
      },
    ],
  };
}

export function code(text: string) {
  const sourceText = dedent(text);
  const lines = splitLines(sourceText);
  let position = null;
  lines.forEach((line, index) => {
    const match = line.match(/--\^/);
    if (match) {
      position = {
        // $FlowDisableNextLine
        column: match.index + 3,
        line: index,
      };
    }
  });

  if (!position) {
    throw new Error('Missing --^ in source');
  }

  return {
    sourceText,
    position,
  };
}
