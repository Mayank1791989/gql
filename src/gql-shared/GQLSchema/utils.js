/* @flow */
import {
  type TypeNode,
  type NamedTypeNode,
  type EnumValueDefinitionNode,
  type FieldDefinitionNode,
  getDirectiveValues,
  GraphQLDeprecatedDirective,
  Kind,
} from 'graphql';

export function getNamedTypeNode(typeNode: TypeNode): NamedTypeNode {
  let namedType = typeNode;
  while (
    namedType.kind === Kind.LIST_TYPE ||
    namedType.kind === Kind.NON_NULL_TYPE
  ) {
    namedType = namedType.type;
  }
  return namedType;
}

/**
 * Given a field or enum value node, returns the string value for the
 * deprecation reason.
 */
export function getDeprecationReason(
  node: EnumValueDefinitionNode | FieldDefinitionNode,
): ?string {
  const deprecated = getDirectiveValues(GraphQLDeprecatedDirective, node);
  return deprecated && (deprecated.reason: any);
}
