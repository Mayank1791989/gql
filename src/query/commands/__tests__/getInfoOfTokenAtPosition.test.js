/* @flow */
import { getInfoOfTokenAtPosition } from '../index';
import { getSchema, code } from '../../../__test-data__/utils';

const schema = getSchema();
const relayQLParser = {
  isRelay: true,
  parser: ['EmbeddedQueryParser', { startTag: 'Relay.QL`', endTag: '`' }],
};

describe('fragments', () => {
  it('on Type', () => {
    const { sourceText, position } = code(`
      const a = Relay.QL\`
        fragment test on Player {
              #-----------^
          id
        }
      \`
    `);
    expect(
      getInfoOfTokenAtPosition(schema, sourceText, position, relayQLParser),
    ).toMatchSnapshot();
  });

  it('field: Should Include description, fieldDefn and output Type', () => {
    const { sourceText, position } = code(`
      const a = Relay.QL\`
        fragment test on NewPlayer {
          friend
          #--^
        }
      \`
    `);
    expect(
      getInfoOfTokenAtPosition(schema, sourceText, position, relayQLParser),
    ).toMatchSnapshot();
  });

  it('field arguments', () => {
    const { sourceText, position } = code(`
      const a = Relay.QL\`
        fragment test on Player {
          image(size
            #----^
        }
      \`
    `);
    expect(
      getInfoOfTokenAtPosition(schema, sourceText, position, relayQLParser),
    ).toMatchSnapshot();
  });
});

describe('directives', () => {
  it('core: name', () => {
    const { sourceText, position } = code(`
      const a = Relay.QL\`
        fragment test on Player {
          image @include(if: true)
            #--------^
        }
      \`
    `);
    expect(
      getInfoOfTokenAtPosition(schema, sourceText, position, relayQLParser),
    ).toMatchSnapshot();
  });

  it('core: args', () => {
    const { sourceText, position } = code(`
      const a = Relay.QL\`
        fragment test on Player {
          image @include(if: true)
                  #-------^
        }
      \`
    `);
    expect(
      getInfoOfTokenAtPosition(schema, sourceText, position, relayQLParser),
    ).toMatchSnapshot();
  });

  it('user defined: name', () => {
    const { sourceText, position } = code(`
      const a = Relay.QL\`
        fragment test on Player {
          image @customDirective(if: true)
            #-------^
        }
      \`
    `);
    expect(
      getInfoOfTokenAtPosition(schema, sourceText, position, relayQLParser),
    ).toMatchSnapshot();
  });

  it('user defined: args', () => {
    const { sourceText, position } = code(`
      const a = Relay.QL\`
        fragment test on Player {
          image @customDirective(if: true)
                          #------^
        }
      \`
    `);
    expect(
      getInfoOfTokenAtPosition(schema, sourceText, position, relayQLParser),
    ).toMatchSnapshot();
  });

  it('unknown: name', () => {
    const { sourceText, position } = code(`
      const a = Relay.QL\`
        fragment test on Player {
          image @xyzeagder(if: true)
            #-------^
        }
      \`
    `);
    expect(
      getInfoOfTokenAtPosition(schema, sourceText, position, relayQLParser),
    ).toEqual(null);
  });

  it('unknown directive: args', () => {
    const { sourceText, position } = code(`
      const a = Relay.QL\`
        fragment test on Player {
          image @xyzeagder(if: true)
                    #-------^
        }
      \`
    `);
    expect(
      getInfoOfTokenAtPosition(schema, sourceText, position, relayQLParser),
    ).toEqual(null);
  });

  it('known directive: unknown args', () => {
    const { sourceText, position } = code(`
      const a = Relay.QL\`
        fragment test on Player {
          image @include(ifs: true)
                  #-------^
        }
      \`
    `);
    expect(
      getInfoOfTokenAtPosition(schema, sourceText, position, relayQLParser),
    ).toEqual(null);
  });

  it('name info', () => {
    const { sourceText, position } = code(`
      const a = Relay.QL\`
        fragment test on Player {
          image @include(if: true)
            #------^
        }
      \`
    `);
    expect(
      getInfoOfTokenAtPosition(schema, sourceText, position, relayQLParser),
    ).toMatchSnapshot();
  });

  it('args', () => {
    const { sourceText, position } = code(`
      const a = Relay.QL\`
        fragment test on Player {
          image @include(if: true)
                   #------^
        }
      \`
    `);
    expect(
      getInfoOfTokenAtPosition(schema, sourceText, position, relayQLParser),
    ).toMatchSnapshot();
  });
});

describe('mutations', () => {
  it('type: include description', () => {
    const { sourceText, position } = code(`
      const a = Relay.QL\`
        mutation {
     #-----^
        }
      \`
    `);
    expect(
      getInfoOfTokenAtPosition(schema, sourceText, position, relayQLParser),
    ).toMatchSnapshot();
  });

  it('field: Include both input and output type', () => {
    const { sourceText, position } = code(`
      const a = Relay.QL\`
        mutation {
          PlayerCreate
          #-----^
        }
      \`
    `);
    expect(
      getInfoOfTokenAtPosition(schema, sourceText, position, relayQLParser),
    ).toMatchSnapshot();
  });

  it('input object fields', () => {
    const { sourceText, position } = code(`
      mutation {
        PlayerCreate(
          input: {
            id: "some_id",
            role:
        #----^
          }
        ) {
          player { id }
        }
      }
    `);
    expect(
      getInfoOfTokenAtPosition(schema, sourceText, position, { parser: 'QueryParser' }),
    ).toMatchSnapshot();
  });
});

describe('query', () => {
  it('query keyword', () => {
    const { sourceText, position } = code(`
      const a = Relay.QL\`
        query Viewer { viewer }
        #--^
      \`
    `);
    expect(
      getInfoOfTokenAtPosition(schema, sourceText, position, relayQLParser),
    ).toMatchSnapshot();
  });
  it('field', () => {
    const { sourceText, position } = code(`
      const a = Relay.QL\`
        query Viewer { viewer }
              #-----------^
      \`
    `);
    expect(
      getInfoOfTokenAtPosition(schema, sourceText, position, relayQLParser),
    ).toMatchSnapshot();
  });
});

describe('meta field __typename', () => {
  it('when inside Object Type', () => {
    const { sourceText, position } = code(`
      const a = Relay.QL\`
        fragment on Viewer {
          __typename
          #----^
        }
      \`
    `);
    expect(
      getInfoOfTokenAtPosition(schema, sourceText, position, relayQLParser),
    ).toMatchSnapshot();
  });
  it('when inside Interface Type', () => {
    const { sourceText, position } = code(`
      const a = Relay.QL\`
        fragment on Node {
          __typename
          #----^
        }
      \`
    `);
    expect(
      getInfoOfTokenAtPosition(schema, sourceText, position, relayQLParser),
    ).toMatchSnapshot();
  });
  it('when inside Union Type', () => {
    const { sourceText, position } = code(`
      const a = Relay.QL\`
        fragment on Entity {
          __typename
          #----^
        }
      \`
    `);
    expect(
      getInfoOfTokenAtPosition(schema, sourceText, position, relayQLParser),
    ).toMatchSnapshot();
  });
});

test('meta field __schema', () => {
  const { sourceText, position } = code(`
    const a = Relay.QL\`
      query Viewer {
        __schema
        #----^
      }
    \`
  `);
  expect(
    getInfoOfTokenAtPosition(schema, sourceText, position, relayQLParser),
  ).toMatchSnapshot();
});

test('meta field __type', () => {
  const { sourceText, position } = code(`
    const a = Relay.QL\`
      query Viewer {
        __type
        #--^
      }
    \`
  `);
  expect(
    getInfoOfTokenAtPosition(schema, sourceText, position, relayQLParser),
  ).toMatchSnapshot();
});

