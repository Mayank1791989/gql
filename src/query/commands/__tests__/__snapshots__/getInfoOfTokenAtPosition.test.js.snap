// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`directives args 1`] = `
Object {
  "contents": Array [
    "# Included when true.
if: Boolean!",
    "# The \`Boolean\` scalar type represents \`true\` or \`false\`.",
  ],
}
`;

exports[`directives core: args 1`] = `
Object {
  "contents": Array [
    "# Included when true.
if: Boolean!",
    "# The \`Boolean\` scalar type represents \`true\` or \`false\`.",
  ],
}
`;

exports[`directives core: name 1`] = `
Object {
  "contents": Array [
    "# Directs the executor to include this field or fragment only when the \`if\` argument is true.
directive @include(
  # Included when true.
  if: Boolean!
)
  on FIELD
   | FRAGMENT_SPREAD
   | INLINE_FRAGMENT",
  ],
}
`;

exports[`directives name info 1`] = `
Object {
  "contents": Array [
    "# Directs the executor to include this field or fragment only when the \`if\` argument is true.
directive @include(
  # Included when true.
  if: Boolean!
)
  on FIELD
   | FRAGMENT_SPREAD
   | INLINE_FRAGMENT",
  ],
}
`;

exports[`directives user defined: args 1`] = `
Object {
  "contents": Array [
    "# description for arg
if: Boolean!",
    "# The \`Boolean\` scalar type represents \`true\` or \`false\`.",
  ],
}
`;

exports[`directives user defined: name 1`] = `
Object {
  "contents": Array [
    "# some custom directive
directive @customDirective(
  # description for arg
  if: Boolean!
)
  on FIELD
   | FRAGMENT_DEFINITION",
  ],
}
`;

exports[`fragments field arguments 1`] = `
Object {
  "contents": Array [
    "(argument) size: Int!",
    "# The \`Int\` scalar type represents non-fractional signed whole numeric values. Int can represent values between -(2^31) and 2^31 - 1. ",
  ],
}
`;

exports[`fragments field: Should Include description, fieldDefn and output Type 1`] = `
Object {
  "contents": Array [
    "(field) friend: Node!",
    "interface Node {
  id: ID!
}",
  ],
}
`;

exports[`fragments on Type 1`] = `
Object {
  "contents": Array [
    "type Player implements Node {
  id: ID!
  name: String!
  image(size: Int!): String!
}",
  ],
}
`;

exports[`meta field __schema 1`] = `
Object {
  "contents": Array [
    "# Access the current type schema of this server.
(meta-field) __schema: __Schema!",
    "# A GraphQL Schema defines the capabilities of a GraphQL server. It exposes all
# available types and directives on the server, as well as the entry points for
# query, mutation, and subscription operations.
type __Schema {
  # A list of all types supported by this server.
  types: [__Type!]!

  # The type that query operations will be rooted at.
  queryType: __Type!

  # If this server supports mutation, the type that mutation operations will be rooted at.
  mutationType: __Type

  # If this server support subscription, the type that subscription operations will be rooted at.
  subscriptionType: __Type

  # A list of all directives supported by this server.
  directives: [__Directive!]!
}",
  ],
}
`;

exports[`meta field __type 1`] = `
Object {
  "contents": Array [
    "# Request the type information of a single type.
(meta-field) __type(name: String!): __Type",
    "# The fundamental unit of any GraphQL Schema is the type. There are many kinds of
# types in GraphQL as represented by the \`__TypeKind\` enum.
#
# Depending on the kind of a type, certain fields describe information about that
# type. Scalar types provide no information beyond a name and description, while
# Enum types provide their values. Object and Interface types provide the fields
# they describe. Abstract types, Union and Interface, provide the Object types
# possible at runtime. List and NonNull types compose other types.
type __Type {
  kind: __TypeKind!
  name: String
  description: String
  fields(includeDeprecated: Boolean = false): [__Field!]
  interfaces: [__Type!]
  possibleTypes: [__Type!]
  enumValues(includeDeprecated: Boolean = false): [__EnumValue!]
  inputFields: [__InputValue!]
  ofType: __Type
}",
  ],
}
`;

exports[`meta field __typename when inside Interface Type 1`] = `
Object {
  "contents": Array [
    "# The name of the current Object type at runtime.
(meta-field) __typename: String!",
    "# The \`String\` scalar type represents textual data, represented as UTF-8 character
# sequences. The String type is most often used by GraphQL to represent free-form
# human-readable text.
scalar String",
  ],
}
`;

exports[`meta field __typename when inside Object Type 1`] = `
Object {
  "contents": Array [
    "# The name of the current Object type at runtime.
(meta-field) __typename: String!",
    "# The \`String\` scalar type represents textual data, represented as UTF-8 character
# sequences. The String type is most often used by GraphQL to represent free-form
# human-readable text.
scalar String",
  ],
}
`;

exports[`meta field __typename when inside Union Type 1`] = `
Object {
  "contents": Array [
    "# The name of the current Object type at runtime.
(meta-field) __typename: String!",
    "# The \`String\` scalar type represents textual data, represented as UTF-8 character
# sequences. The String type is most often used by GraphQL to represent free-form
# human-readable text.
scalar String",
  ],
}
`;

exports[`mutations field: Include both input and output type 1`] = `
Object {
  "contents": Array [
    "# Player Create contains all allowed mutations
(field) PlayerCreate(input: PlayerCreateInput!): PlayerCreatePayload",
    "input PlayerCreateInput {
  id: ID!
  name: String!
  role: Role!
}",
    "type PlayerCreatePayload {
  player: Player!
}",
  ],
}
`;

exports[`mutations input object fields 1`] = `
Object {
  "contents": Array [
    "(field) role: Role!",
    "enum Role {
  roleA
  roleB
  roleC
}",
  ],
}
`;

exports[`mutations type: include description 1`] = `
Object {
  "contents": Array [
    "# Mutation contains all allowed mutations
type Mutation {
  # Player Create contains all allowed mutations
  PlayerCreate(input: PlayerCreateInput!): PlayerCreatePayload
}",
  ],
}
`;

exports[`query field 1`] = `
Object {
  "contents": Array [
    "(field) viewer: Viewer!",
    "type Viewer {
  me: Player!
  xme: XPlayer! # validation error
}",
  ],
}
`;

exports[`query query keyword 1`] = `
Object {
  "contents": Array [
    "# Query is the root query object
type Query {
  node(id: String!): Node
  nodes(ids: [String!]): [Node]
  viewer: Viewer!
}",
  ],
}
`;
